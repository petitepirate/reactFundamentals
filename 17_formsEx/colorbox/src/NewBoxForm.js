import React from 'react';
import './NewBoxForm.css'

function NewBoxForm({ formData, submitHandler, changeHandler}) {
  return (
    <form onSubmit={submitHandler}>
      <label htmlFor="width">Box Width (in pixels):</label>
      <input
        type="number"
        id="width"
        name="width"
        value={formData.width}
        onChange={changeHandler}
      />
      <label htmlFor="height">Box Height (in pixels):</label>
      <input
        type="number"
        id="height"
        name="height"
        value={formData.height}
        onChange={changeHandler}
      />
      <label htmlFor="boxBackgroundColor">Box Background Color(css color name):</label>
      <input
        type="text"
        id="boxBackgroundColor"
        name="boxBackgroundColor"
        value={formData.boxBackgroundColor}
        onChange={changeHandler}
      />
      <button>Create Box</button>
    </form>
  )
}

export default NewBoxForm
