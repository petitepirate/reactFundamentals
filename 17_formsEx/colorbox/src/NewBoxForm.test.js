import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import NewBoxForm from './NewBoxForm';

describe('NewBoxForm Component', () => {
  const mockChangeHandler = jest.fn();
  const mockSubmitHandler = jest.fn();
  const intialFormValue = {
    width: '', 
    height: '', 
    boxBackgroundColor: ''
  }
  const changedFormData = {...intialFormValue, width: '100'};

  it('renders form without crashing', () => {
    render(<NewBoxForm 
      formData={intialFormValue} 
      changeHandler={mockChangeHandler} 
      submitHandler={mockSubmitHandler} 
      />)
  }) 

  it('matches snapshot', () => {
    const { asFragment } = render(<NewBoxForm 
      formData={intialFormValue}
      changeHandler={mockChangeHandler} 
      submitHandler={mockSubmitHandler}
      />);
      expect(asFragment()).toMatchSnapshot();
  })

  it('tests form operation', () => {
    const { getByLabelText, rerender, queryByText } = render(<NewBoxForm 
      formData={intialFormValue}
      changeHandler={mockChangeHandler} 
      submitHandler={mockSubmitHandler}
      />);
    
      const widthInput = getByLabelText('Box Width (in pixels):');
      const heightInput = getByLabelText('Box Height (in pixels):');
      const boxBackgroundColorInput = getByLabelText('Box Background Color(css color name):');
      const addBoxBtn = queryByText('Create Box')
      
      expect(widthInput.value).toBe('');
      expect(heightInput.value).toBe('');
      expect(boxBackgroundColorInput.value).toBe('');

      fireEvent.change(widthInput, { target: { value: '100'}});

      rerender(<NewBoxForm 
        formData={changedFormData}
        changeHandler={mockChangeHandler} 
        submitHandler={mockSubmitHandler}
        />);

      expect(mockChangeHandler).toBeCalledTimes(1);
      expect(widthInput.value).toBe('100');

      fireEvent.click(addBoxBtn);

      expect(mockSubmitHandler).toBeCalledTimes(1);
  }) 

})
