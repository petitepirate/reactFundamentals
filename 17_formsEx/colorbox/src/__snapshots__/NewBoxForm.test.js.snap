// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewBoxForm Component matches snapshot 1`] = `
<DocumentFragment>
  <form>
    <label
      for="width"
    >
      Box Width (in pixels):
    </label>
    <input
      id="width"
      name="width"
      type="number"
      value=""
    />
    <label
      for="height"
    >
      Box Height (in pixels):
    </label>
    <input
      id="height"
      name="height"
      type="number"
      value=""
    />
    <label
      for="boxBackgroundColor"
    >
      Box Background Color(css color name):
    </label>
    <input
      id="boxBackgroundColor"
      name="boxBackgroundColor"
      type="text"
      value=""
    />
    <button>
      Create Box
    </button>
  </form>
</DocumentFragment>
`;
